File: src/main/java/com/github/jaksonlin/pitestintellij/context/CaseCheckContext.java
package com.github.jaksonlin.pitestintellij.context;

import com.github.jaksonlin.pitestintellij.annotations.AnnotationParser;
import com.github.jaksonlin.pitestintellij.annotations.AnnotationSchema;
import com.github.jaksonlin.pitestintellij.services.AnnotationConfigService;
import com.github.jaksonlin.pitestintellij.services.InvalidAssertionConfigService;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiMethod;

import java.util.List;

public class CaseCheckContext {
    private final PsiClass psiClass;
    private final PsiMethod psiMethod;
    private final AnnotationSchema schema;
    private final AnnotationParser parser;
    private final List<String> invalidAssertions;

    public CaseCheckContext(PsiClass psiClass, PsiMethod psiMethod, AnnotationSchema schema, AnnotationParser parser, List<String> invalidAssertions) {
        this.psiClass = psiClass;
        this.psiMethod = psiMethod;
        this.schema = schema;
        this.parser = parser;
        this.invalidAssertions = invalidAssertions;
    }

    public PsiClass getPsiClass() {
        return psiClass;
    }

    public PsiMethod getPsiMethod() {
        return psiMethod;
    }

    public AnnotationSchema getSchema() {
        return schema;
    }

    public AnnotationParser getParser() {
        return parser;
    }

    public CaseCheckContext copy(PsiMethod newPsiMethod) {
        return new CaseCheckContext(psiClass, newPsiMethod, schema, parser, invalidAssertions);
    }

    public static CaseCheckContext create(PsiMethod psiMethod, PsiClass psiClass) {
        AnnotationConfigService configService = ApplicationManager.getApplication().getService(AnnotationConfigService.class);
        AnnotationSchema schema = configService.getSchema();
        InvalidAssertionConfigService invalidAssertionConfigService = ApplicationManager.getApplication().getService(InvalidAssertionConfigService.class);
        return new CaseCheckContext(
                psiClass,
                psiMethod,
                schema,
                new AnnotationParser(schema),
                invalidAssertionConfigService.getInvalidAssertions()
        );
    }
}

File: src/main/kotlin/com/github/jaksonlin/pitestintellij/context/CaseCheckContext.kt
package com.github.jaksonlin.pitestintellij.context

import AnnotationConfigService
import com.github.jaksonlin.pitestintellij.actions.RunCaseAnnoationCheckAction
import com.github.jaksonlin.pitestintellij.annotations.AnnotationParser
import com.github.jaksonlin.pitestintellij.annotations.AnnotationSchema
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiMethod

data class CaseCheckContext(
    val psiMethod: PsiMethod,
    val schema: AnnotationSchema,
    val parser: AnnotationParser
) {
    companion object {
        fun create(psiMethod: PsiMethod): CaseCheckContext? {
            val configService = service<AnnotationConfigService>()
            val schema = configService.getSchema()
            return CaseCheckContext(
                psiMethod = psiMethod,
                schema = schema,
                parser = AnnotationParser(schema)
            )
        }
    }
}